nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -202,7 +202,7 @@ in
 
         environment = {
           LEMMY_CONFIG_LOCATION = "${settingsFormat.generate "config.hjson" cfg.settings}";
-          LEMMY_DATABASE_URL = mkIf (cfg.database.uri != null) cfg.database.uri;
+          LEMMY_DATABASE_URL = if cfg.database.uri != null then cfg.database.uri else (mkIf (cfg.database.createLocally) "postgres:///lemmy?host=/run/postgresql&user=lemmy");
```

Comment:
how about putting the mkIf after line 105, this is where the defaults are. (it would be more consistent to keep the defaults together).

Reply:
Not sure where exactly you are referring to, I don't see any other references to LEMMY_DATABASE_URL

Reply:
The idea is not to modify the environment variable directly but the database.uri
The line i referenced sets all the default parameters for the database. We could add another line there defining "uri"
Then this environment variable gets defined automatically

Let me know if it makes sense.

Reply:
It does, at least I think I understand now.

Will make the changes in a bit.

Reply:
```diff
database = mapAttrs (name: mkDefault) {
          user = "lemmy";
          host = "/run/postgresql";
          port = 5432;
          database = "lemmy";
          pool_size = 5;
+          uri = if cfg.database.uri != null then cfg.database.uri else (mkIf (cfg.database.createLocally) "postgres:///lemmy?host=/run/postgresql&user=lemmy");
        };

Reply:
that is the right place for the change.
The change is almost correct IMO.
All of these settings are mapped to mkDefault, supposed to act as a default. So if uri is defined, it will override the value we use here. Therefore we don't need to check for cfg.database.uri. We only need the mkIf part.
Also for the actual uri itself, you shouldn't need the host and user, just `postgres:///lemmy` should be enough.

Reply:
The problem with this approach is that the parent attribute (lemmy.settings) is then parsed to write the rjson file, which doesn't contain a line for the database connection.

Reply:
Ah, so you tried and then the parsing failed on startup, is that correct ? If thats the case, then youre right, where you have it is most likely the best place for it.

Also did you try without specifying the host and the user, it should work (the host is definitely the default, and the user should be omittable too)

Reply:
I'm not sure what the problem is here

Reply:
@happysalada I think you're confusing `services.lemmy.database.uri` and `services.lemmy.settings.database.uri`
The latter will put the uri in the json settings file, which does **not** get picked up by `lemmy-server`.

Setting the former is what is required.

Take a look at this:

![image](https://github.com/NixOS/nixpkgs/assets/30625554/d6bf3e8d-6124-4688-a02a-4380bc535281)
![image](https://github.com/NixOS/nixpkgs/assets/30625554/c8646cb0-7172-4b95-8109-824f87514689)


and now, with the current implementation of this PR:
![image](https://github.com/NixOS/nixpkgs/assets/30625554/644eb3e3-e6d7-4d5f-a881-f28b5ff4d496)
(no `database.uri` set)
![image](https://github.com/NixOS/nixpkgs/assets/30625554/b6d2b837-40c9-4c55-8f06-703606c4dfaf)
site works fine

Reply:
Also, for using just `postgres:///lemmy` - it doesn't work for me :(
I have to use the full URL.
![image](https://github.com/NixOS/nixpkgs/assets/30625554/412c841e-962b-4301-ba2d-fc7ba9e43751)

Reply:
@n3oney thanks for the clarification!
The current implementation does work, there is no doubt about that, I just think it's a bit confusing to have defaults in 2 places.
I had indeed missed that database.uri and settings.database.uri are very different.
how about adding this line 84 then ?
```
      services.lemmy.database.uri = mkIf (cfg.databse.createLocally) (mkDefault "postgres:///lemmy?host=/run/postgresql&user=lemmy");
```

The intent is just to try to keep the defaults together so that if someone wants to debug, they don't need to look in too many places.

Sorry for my late answer on that, I was travelling.

btw, someone is making a PR on top of this one to keep the secrets out of the nix store. Let me know if you are interested in reviewing that one too.